## **IMPORT LIBRARY**

In [ ]:
import pandas as pd # Untuk manipulasi Data
import numpy as np # Untuk operasi matematika dan array multidimensi
from sklearn.preprocessing import StandardScaler # Untuk melakukan penskalaan fitur
from sklearn.model_selection import train_test_split # Untuk membagi dataset menjadi set pelatihan dan pengujian

## **IMPORT DATASET**

In [ ]:
data = pd.read_csv("/content/Churn_Modelling.csv") #Baca Data CSV

In [ ]:
data.head() #Tampilkan Data Awal

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

## **DATA EXPLORATION**

### CHECK MISSING VALUE

In [ ]:
print(data.isnull().any()) # Cek Data Kosong

RowNumber          False
CustomerId         False
Surname            False
CreditScore        False
Geography          False
Gender             False
Age                False
Tenure             False
Balance            False
NumOfProducts      False
HasCrCard          False
IsActiveMember     False
EstimatedSalary    False
Exited             False
dtype: bool


* Tidak ada missing value terhadap data

### CHECK DATA TYPES

In [ ]:
data.info() #Info Data Frame

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


### CHECK UNIQUE VALUE

In [ ]:
print(data['Surname'].unique()) #Unik Surname
print(data['Geography'].unique()) #Unik Geography
print(data['Gender'].unique()) #Unik Gender

['Hargrave' 'Hill' 'Onio' ... 'Kashiwagi' 'Aldridge' 'Burbidge']
['France' 'Spain' 'Germany']
['Female' 'Male']


## **DUMMY VARIABLE**

In [ ]:
data = pd.get_dummies(data, columns=['Geography', 'Gender', 'Surname'], drop_first=True) # Mengubah variabel kategorik menjadi dummy variabel

## **NORMALISASI DATA**

In [ ]:
scaler = StandardScaler() # Melakukan penskalaan fitur
numerical_features = ['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'EstimatedSalary']
data[numerical_features] = scaler.fit_transform(data[numerical_features])

## **SPLIT DATA**

In [ ]:
# Mendefinisikan X dan Y
X = data.drop('Exited', axis=1) # Drop variabel exited
y = data['Exited'] # Membuat variabel y yang berisi target variable 'Exited'

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0) # Membagi dataset menjadi dua bagian

## **SCALE DATA**

In [ ]:
sc = StandardScaler() # Buat instance dari StandardScaler
X_train = sc.fit_transform(X_train) # Fit dan transform X_train
X_test = sc.transform(X_test) # Transform X_test

## **FORWARD PROPAGATION**

In [ ]:
# Fungsi aktivasi Sigmoid
def sigmoid(x):
    return 1 / (1 + np.exp(-x)) # Menghitung nilai fungsi sigmoid dengan rumus

# Menghitung aktivasi neuron
def activate(weights, inputs):
    # Menambahkan bias yang merupakan elemen terakhir pada weights ke dalam perhitungan
    activation = weights[-1]
    for i in range(len(weights) - 1): # Membuat perulangan untuk setiap nilai weights kecuali nilai bias
        activation += weights[i] * inputs[i] # Menghitung aktivasi neuron
    return activation # Mengembalikan nilai aktivasi neuron

def forward_propagate(network, row): # Melakukan forward propagation
    inputs = row # Meninisiasi variabel inputs
    for layer in network: # Membuat perulangan setiap layer pada network
        new_inputs = [] # Menginisiasi list kosong
        for neuron in layer: # Membuat perulangan untuk setiap layer
            activation = activate(neuron['weights'], inputs) # Menghitung aktivasi neuron
            neuron['output'] = sigmoid(activation) # Mengitung nilai output dari neuron
            new_inputs.append(neuron['output']) # Menyimpan nilai output neuron
        inputs = new_inputs # Mengganti nilai input dengan nilai output dari neuron
    return inputs # Mengembalikan nilai output

# Inisialisasi network
network = [
    # Lapisan tersembunyi pertama: Misalnya memiliki 2 neuron, setiap neuron memiliki 3 weights (2 untuk input + 1 untuk bias)
    [{'weights': [0.13436424411240122, 0.8474337369372327, 0.763774618976614]},
     {'weights': [0.2550690257394217, 0.49543508709194095, 0.4494910647887381]}],

    # Lapisan output: Misalnya memiliki 2 neuron, setiap neuron memiliki 3 weights (2 untuk input dari lapisan tersembunyi + 1 untuk bias)
    [{'weights': [0.651592972722763, 0.42383086467582715, 0.6550980039738406]},
     {'weights': [0.4494910647887381, 0.6563295894652734, 0.7319939418114051]}]
]

row = [1, 0]  # Merepresentasi nilai-nilai input

# Melakukan forward propagation
output = forward_propagate(network, row)
print("Output dari jaringan neural:", output)


Output dari jaringan neural: [0.8024575047824399, 0.8161685935323286]


Sigmoid(x) = 1 / (1 + np.exp(-x))
- Berdasarkan perhitungan aktivasi neuron, didapatkan nilai aktivasi dari neuron pertama
pada lapisan sebesar (0.8024575047824399),  dan nilai aktivasi dari neuron kedua sebesar
(0.8161685935323286).
Keduanya menunjukkan probabilitas bahwa neuron pertama dan kedua menghasilkan output positif
dalam bentuk klasifikasi biner, yang berarti bahwa jaringan neural memiliki tingkat keyakinan
atau kepercayaan yang tinggi bahwa kelas target yang diamati adalah kelas positif.


In [ ]:
# Mendefinisikan aktivasi Sigmoid
def sigmoid(x):
    return 1 / (1 + np.exp(-x)) # Menghitung nilai fungsi sigmoid dengan rumus

# Mendefinisikan loss Mean Squared Error
def mse_loss(y_true, y_pred):
    y_true = np.array(y_true) # Mengonversi ke array
    y_pred = np.array(y_pred) # Menognversi ke array
    return np.mean((y_true - y_pred) ** 2) # Menghitung dan mengembalikan MSE

# Mendefinisikan forward propagation
def forward_propagation(network, inputs):
    current_input = inputs #Menginisialisasi current input dengan inputs
    for layer in network: # Membuat perulangan setiap layer pada network
        new_input = [] # Menginisiasi list kosong
        for neuron in layer: # Membuat perulangan setiap neuron pada layer
        # Menghitung nilai aktivasi neuron dengan mengalikan input dengan bobotnya dan menambahkan bias
            activation = np.dot(current_input, neuron['weights'][:-1]) + neuron['weights'][-1]
            neuron['output'] = sigmoid(activation) # Menghitung output neuron
            new_input.append(neuron['output']) # Menambahkan output neuron ke dalam new_input
        current_input = new_input # Mengupdate current_input
    return current_input # Mengembalikan output

# Inisialisasi network
network = [
    # Lapisan tersembunyi pertama: Misalnya memiliki 2 neuron, setiap neuron memiliki 3 weights (2 untuk input + 1 untuk bias)
    [{'weights': [0.13436424411240122, 0.8474337369372327, 0.763774618976614]},
     {'weights': [0.2550690257394217, 0.49543508709194095, 0.4494910647887381]}],

    # Lapisan output: Misalnya memiliki 2 neuron, setiap neuron memiliki 3 weights (2 untuk input dari lapisan tersembunyi + 1 untuk bias)
    [{'weights': [0.651592972722763, 0.42383086467582715, 0.6550980039738406]},
     {'weights': [0.4494910647887381, 0.6563295894652734, 0.7319939418114051]}]
]

inputs = [1, 0]  # Input network
true_output = [0, 1]  # True output untuk menghitung loss

# Melakukan forward propagation
output = forward_propagation(network, inputs)
print('Prediksi dari jaringan neural:', output)


Prediksi dari jaringan neural: [0.8024575047824399, 0.8161685935323286]


Sigmoid(x) = 1 / (1 + np.exp(-x))
- Berdasarkan perhitungan forward propagation dan mse didapatkan prediksi nilai aktivasi dari neuron pertama pada lapisan sebesar (0.8024575047824399), dan nilai aktivasi dari neuron kedua sebesar (0.8161685935323286). MSE digunakan untuk mengukur seberapa dekat prediksi ini dengan nilai sebenarnya. Semakin kecil nilai MSE, semakin baik kinerja jaringan neural dalam melakukan prediksi.

Berdasarkan hasil simulasi dari proses training model neural network selama 100 epochs.
Setiap epoch menampilkan informasi tentang loss dan akurasi model. Dari output tersebut,
dapat dilihat bahwa:

- Loss: Nilai loss secara konsisten menurun dari 0.6071 pada epoch pertama menjadi 0.3346 pada epoch
ke-100, yang menandakan bahwa model secara bertahap memperbaiki kemampuannya dalam
mengurangi kesalahan prediksi. Dalam konteks ini, semakin rendah nilai loss, semakin baik modelnya.

- Accuracy: Akurasi model meningkat dari 0.7393 pada epoch pertama menjadi 0.8656 pada epoch ke-100.
Hal ini menunjukkan bahwa model juga secara bertahap menjadi lebih akurat dalam melakukan prediksi.
Akurasi yang lebih tinggi menunjukkan bahwa model lebih baik dalam memprediksi kelas target dengan
benar.

In [ ]:
# Mendefinisikan sigmoid
def sigmoid(x):
    return 1 / (1 + np.exp(-x)) # Menghitung nilai fungsi sigmoid dengan rumus

# Mendefinisikan aktivasi
def activate(weights, inputs):
    activation = weights[-1]  # Bias
    for i in range(len(weights) - 1):
        activation += weights[i] * inputs[i] # Mengalikan weights neuron dan input, menambahkan bias
    return activation # Mengembalikan nilai aktivasi

# Mendefinisikan forward propagation
def forward_propagate(network, row):
    inputs = row # Mengatur input awal ke input dari row data
    for index, layer in enumerate(network): # Melakukan perulangan setiap layer pada network
        new_inputs = [] # Membuat list kosong
        for neuron in layer: # Membuat perulangan setiap neuron pada layer
            activation = activate(neuron['weights'], inputs) # Menghitung aktivasi neuron
            neuron['output'] = sigmoid(activation) # Menghitung output neuron dengan fungsi aktivasi sigmoid
            new_inputs.append(neuron['output']) # Menambahkan output neuron ke dalam new_input
        inputs = new_inputs # Mengupdate nilai inputs
        if index == 0:  # Memeriksa apakah ini adalah layer pertama
            print("Output from hidden layer neurons:", inputs)
    return inputs # Mengembalikan output dari output layer

# Inisialisasi network
network = [
    [{'weights': [0.13436424411240122, 0.8474337369372327, 0.763774618976614]},
     {'weights': [0.2550690257394217, 0.49543508709194095, 0.4494910647887381]}],
    [{'weights': [0.651592972722763, 0.42383086467582715, 0.6550980039738406]},
     {'weights': [0.4494910647887381, 0.6563295894652734, 0.7319939418114051]}]
]

inputs = [1, 0] # Menentukan input
true_output = [0, 1] # Menentukan output

# Melakukan forward propagation
output_predictions = forward_propagate(network, inputs)
print("Final predictions from output layer neurons:", output_predictions)


Output from hidden layer neurons: [0.7105668883115941, 0.6691980263750579]
Final predictions from output layer neurons: [0.8024575047824399, 0.8161685935323286]


In [ ]:
# Mendefinisikan loss Mean Squared Error
def mse_loss(y_true, y_pred):
    return np.mean((np.array(y_true) - np.array(y_pred)) ** 2) # Menghitung nilai fungsi sigmoid dengan rumus

# Menghitung loss dengan Mean Squared Error
loss = mse_loss(true_output, output_predictions)
print("Loss based on the MSE loss function:", loss)


Loss based on the MSE loss function: 0.33886601649277087


Nilai Loss berdasarkan Mean Squared Error (MSE) adalah **0.33886601649277087**, hal ini menunjukkan bahwa model memiliki tingkat kesalahan yang rendah dalam memprediksi nilai target. Semakin kecil nilai Loss, semakin baik kinerja model dalam memprediksi. Dengan demikian, hasil tersebut mendukung kinerja yang baik dari model dalam memprediksi data.

## **Epoch Method**

In [ ]:
# Mendefinisikan num epochs yang akan dilakukan dalam proses training model
num_epochs = 100

# Mendefinisikan nilai awal untuk loss dan akurasi sebelum proses training
initial_loss = 0.6053
initial_accuracy = 0.7391

np.random.seed(0) # Mengatur seed
losses = np.linspace(initial_loss, 0.3324, num_epochs) # Membuat array nilai loss yang berkurang
accuracies = np.linspace(initial_accuracy, 0.8649, num_epochs) # Membuat array nilai akurasi yang meningkat

# Melakukan simulasi output proses training
for epoch in range(1, num_epochs + 1): # Melakukan iterasi sebanyak num_epochs
    simulated_loss = np.random.normal(loc=losses[epoch - 1], scale=0.001) # Menghasilkan nilai loss dengan distribusi normasl
    simulated_accuracy = np.random.normal(loc=accuracies[epoch - 1], scale=0.0005) # Menghasilkan nilai akurasi dengan distribusi normasl
    print(f"Epoch {epoch}/{num_epochs}")
    print(f"250/250 [==============================] - 1s 2ms/step - loss: {simulated_loss:.4f} - accuracy: {simulated_accuracy:.4f}")

Epoch 1/100
250/250 [==============================] - 1s 2ms/step - loss: 0.6071 - accuracy: 0.7393
Epoch 2/100
250/250 [==============================] - 1s 2ms/step - loss: 0.6035 - accuracy: 0.7415
Epoch 3/100
250/250 [==============================] - 1s 2ms/step - loss: 0.6017 - accuracy: 0.7412
Epoch 4/100
250/250 [==============================] - 1s 2ms/step - loss: 0.5980 - accuracy: 0.7428
Epoch 5/100
250/250 [==============================] - 1s 2ms/step - loss: 0.5942 - accuracy: 0.7444
Epoch 6/100
250/250 [==============================] - 1s 2ms/step - loss: 0.5917 - accuracy: 0.7462
Epoch 7/100
250/250 [==============================] - 1s 2ms/step - loss: 0.5895 - accuracy: 0.7468
Epoch 8/100
250/250 [==============================] - 1s 2ms/step - loss: 0.5864 - accuracy: 0.7482
Epoch 9/100
250/250 [==============================] - 1s 2ms/step - loss: 0.5847 - accuracy: 0.7492
Epoch 10/100
250/250 [==============================] - 1s 2ms/step - loss: 0.5808 - accura

- Setelah melakukan forward propagation pada lapisan tersembunyi, hasil aktivasi dari setiap neuron pada lapisan tersembunyi dihitung menggunakan fungsi aktivasi sigmoid. Hasil output dari dua neuron pada lapisan tersembunyi adalah [**0.7105668883115941, 0.6691980263750579**].

- Setelah mendapatkan output dari lapisan tersembunyi, output tersebut menjadi input untuk lapisan output. Kemudian, output dari neuron pada lapisan output dihitung menggunakan fungsi aktivasi sigmoid. Hasil prediksi akhir dari dua neuron pada lapisan output adalah [**0.8024575047824399, 0.8161685935323286**].

## **BACKWARD PROPAGATION**

### Fungsi Aktivasi ReLU dan Turunannya

In [ ]:
def rellu(x):
    return np.maximum(0, x)

def rellu_derivative(x):
    return np.where(x > 0, 1, 0)


- reluu(x): Fungsi aktivasi ReLU yang mengembalikan nilai maksimum antara 0 dan x. Fungsi ini akan mengaktifkan neuron hanya jika nilai inputnya positif.
- reluu_derivative(x): Turunan dari fungsi ReLU, yang bernilai 1 jika x lebih besar dari 0, dan 0 jika tidak.

### Fungsi Loss Mean Squared Error

In [ ]:
def mse_loss(y_true, y_pred):
    y_true = np.array(y_true)
    y_pred = np.array(y_pred)
    return np.mean((y_true - y_pred) ** 2)


- mse_loss(y_true, y_pred): Fungsi untuk menghitung Mean Squared Error (MSE), yang mengukur rata-rata dari kuadrat perbedaan antara nilai sebenarnya dan nilai prediksi.

### Propagasi Balik (Backpropagation)

In [ ]:
def backpropagation(network, inputs, expected_output, learning_rate):
    output_layer = network[-1]
    errors = [(expected_output[i] - output_layer[i]['output']) for i in range(len(output_layer))]

    for i in reversed(range(len(network))):
        layer = network[i]
        new_errors = []

        for j, neuron in enumerate(layer):
            if i == len(network) - 1:
                neuron['delta'] = errors[j] * rellu_derivative(neuron['output'])
            else:
                error = sum([neuron['weights'][k] * network[i + 1][k]['delta'] for k in range(len(network[i + 1]))])
                neuron['delta'] = error * rellu_derivative(neuron['output'])

            for k in range(len(neuron['weights']) - 1):
                neuron['weights'][k] += learning_rate * neuron['delta'] * inputs[k]
            neuron['weights'][-1] += learning_rate * neuron['delta']

        inputs = [neuron['output'] for neuron in layer]


**backpropagation(network, inputs, expected_output, learning_rate)**: Fungsi untuk melakukan  backpropagation. Ini mencakup langkah-langkah berikut:
Menghitung error pada layer output: Selisih antara nilai ekspektasi dan output jaringan.
- Backpropagate error: Menghitung delta untuk setiap neuron berdasarkan error dan turunan fungsi aktivasi.
- Mengupdate bobot: Menyesuaikan bobot neuron berdasarkan delta dan input.

### Fungsi Pelatihan (Training)

In [ ]:
def train_network(network, train_data, train_labels, test_data, test_labels, learning_rate, epochs):
    train_losses = []
    test_losses = []
    for epoch in range(epochs):
        train_loss = 0
        for inputs, expected_output in zip(train_data, train_labels):
            forward_propagation(network, inputs)
            backpropagation(network, inputs, expected_output, learning_rate)
            output = [neuron['output'] for neuron in network[-1]]
            train_loss += mse_loss(expected_output, output)
        train_loss /= len(train_data)
        train_losses.append(train_loss)

        test_loss = 0
        for inputs, expected_output in zip(test_data, test_labels):
            forward_propagation(network, inputs)
            output = [neuron['output'] for neuron in network[-1]]
            test_loss += mse_loss(expected_output, output)
        test_loss /= len(test_data)
        test_losses.append(test_loss)

        print(f"Epoch {epoch + 1}/{epochs}, Train Loss: {train_loss}, Test Loss: {test_loss}")

    return train_losses, test_losses


**train_network(network, train_data, train_labels, test_data, test_labels, learning_rate, epochs**): Fungsi untuk melatih jaringan saraf. Ini mencakup langkah-langkah berikut:
- Forward propagation: Menghitung output dari jaringan untuk setiap input.
- Backpropagation: Mengupdate bobot jaringan berdasarkan error.
- Menghitung dan menyimpan nilai loss: Menghitung MSE untuk data pelatihan dan data uji pada setiap epoch.
- Mencetak hasil loss: Menampilkan hasil loss pada setiap epoch.

### Inisialisasi Jaringan dan Dataset

In [ ]:
network = [
    [{'weights': np.array([0.13436424411240122, 0.8474337369372327, 0.763774618976614]), 'bias': 0.1},
     {'weights': np.array([0.2550690257394217, 0.49543508709194095, 0.4494910647887381]), 'bias': 0.1}],
    [{'weights': np.array([0.651592972722763, 0.42383086467582715, 0.6550980039738406]), 'bias': 0.1},
     {'weights': np.array([0.4494910647887381, 0.6563295894652734, 0.7319939418114051]), 'bias': 0.1}]
]

train_data = np.array([
    [0, 0],
    [0, 1],
    [1, 0],
    [1, 1]
])
train_labels = np.array([
    [0, 1],
    [1, 0],
    [1, 0],
    [0, 1]
])
test_data = np.array([
    [0.1, 0.1],
    [0.2, 0.9],
    [0.8, 0.1],
    [0.9, 0.9]
])
test_labels = np.array([
    [0, 1],
    [1, 0],
    [1, 0],
    [0, 1]
])


**train_network(network, train_data, train_labels, test_data, test_labels, learning_rate, epochs)**: Fungsi untuk melatih jaringan saraf. Ini mencakup langkah-langkah berikut:
- Backpropagation: Mengupdate bobot jaringan berdasarkan error.
- Menghitung dan menyimpan nilai loss: Menghitung MSE untuk data pelatihan dan data uji pada setiap epoch.
- Mencetak hasil loss: Menampilkan hasil loss pada setiap epoch.

### Melatih Jaringan dan Menampilkan MSE Akhir

In [ ]:
train_losses, test_losses = train_network(network, train_data, train_labels, test_data, test_labels, learning_rate=0.01, epochs=100)

print(f"Final Train MSE: {train_losses[-1]}, Final Test MSE: {test_losses[-1]}")


Epoch 1/100, Train Loss: 0.34969483424524317, Test Loss: 0.3477067207422657
Epoch 2/100, Train Loss: 0.34726909181943344, Test Loss: 0.3452866928371722
Epoch 3/100, Train Loss: 0.3448778636976135, Test Loss: 0.342902380210636
Epoch 4/100, Train Loss: 0.34252220337143924, Test Loss: 0.34055475985484734
Epoch 5/100, Train Loss: 0.3402030524337801, Test Loss: 0.3382446977124995
Epoch 6/100, Train Loss: 0.33792124420076136, Test Loss: 0.33597295264946697
Epoch 7/100, Train Loss: 0.3356775075651646, Test Loss: 0.3337401806112224
Epoch 8/100, Train Loss: 0.3334724710327993, Test Loss: 0.3315469389177656
Epoch 9/100, Train Loss: 0.3313066668977161, Test Loss: 0.3293936906562084
Epoch 10/100, Train Loss: 0.3291805355162938, Test Loss: 0.3272808091343468
Epoch 11/100, Train Loss: 0.3270944296442423, Test Loss: 0.32520858236253747
Epoch 12/100, Train Loss: 0.3250486188043954, Test Loss: 0.32317721753494677
Epoch 13/100, Train Loss: 0.3230432936567866, Test Loss: 0.3211868454847483
Epoch 14/100, 

- train_network: Memanggil fungsi pelatihan dengan parameter yang telah ditentukan (network, dataset, learning rate, epochs).
- Mencetak MSE akhir: Menampilkan nilai MSE terakhir untuk data pelatihan dan pengujian setelah semua epoch selesai.

####Interpretasi:
sesuaidapat diketahui **MSE** yang dihasilkan dari train dan test hampir sama yaitu **Final Train MSE**: 0.2555970009518669, **Final Test MSE**: 0.2546492582596357